core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -100,40 +99,27 @@ abstract class DelayedOperation(override val delayMs: Long,
   def tryComplete(): Boolean
 
   /**
-   * Thread-safe variant of tryComplete() that attempts completion only if the lock can be acquired
-   * without blocking.
+   * Thread-safe variant of tryComplete() that attempts completion after it succeed to hold the lock.
    *
-   * If threadA acquires the lock and performs the check for completion before completion criteria is met
-   * and threadB satisfies the completion criteria, but fails to acquire the lock because threadA has not
-   * yet released the lock, we need to ensure that completion is attempted again without blocking threadA
-   * or threadB. `tryCompletePending` is set by threadB when it fails to acquire the lock and at least one
-   * of threadA or threadB will attempt completion of the operation if this flag is set. This ensures that
-   * every invocation of `maybeTryComplete` is followed by at least one invocation of `tryComplete` until
-   * the operation is actually completed.
+   * There is a long story about using "lock" or "tryLock".
+   *
+   * 1) using lock - There were a lot of cases that a thread holds a group lock and then it tries to hold more group
+   * locks to complete delayed requests. Unfortunately, the scenario causes deadlock and so we had introduced the
+   * "tryLock" to avoid deadlock.
+   *
+   * 2) using tryLock -  However, the "tryLock" causes another issue that the delayed requests may be into
+   * oblivion if the thread, which should complete the delayed requests, fails to get the lock.
+   *
+   * Now, we go back to use "lock" and make sure the thread which tries to complete delayed requests does NOT hold lock.
+   * We introduces a flag "completeDelayedRequests" to ReplicaManager.appendRecords() (and related methods).
+   *
+   * 1) If "completeDelayedRequests" is true, the delayed requests may be completed as much as possible.
+   * 2) If "completeDelayedRequests" is false, the delayed requests are NOT completed and the watch key (for example,
+   *    group key) are returned. Callers can complete the delayed requests manually.
```

Comment:
Callers can complete the delayed requests manually => Callers can complete the delayed requests after releasing any conflicting lock.